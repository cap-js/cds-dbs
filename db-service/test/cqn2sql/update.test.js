'use strict'
const cds = require('@sap/cds/lib')
const cqn2sql = require('../../lib/cqn2sql')

beforeAll(async () => {
  cds.model = await cds.load(__dirname + '/testModel').then(cds.linked)
})
describe('.update', () => {
  test('test with entity of type string', () => {
    const cqnUpdate = {
      UPDATE: {
        entity: { ref: ['Foo2'] },
        with: { ID: { val: 1 }, name: { val: "'asd'" }, a: { val: 2 } },
      },
    }

    const { sql, values } = cqn2sql(cqnUpdate)
    expect({ sql, values }).toMatchSnapshot()
  })

  test('test with entity of type string and where clause', () => {
    const cqnUpdate = {
      UPDATE: {
        entity: { ref: ['Foo2'] },
        with: { ID: { val: 1 }, name: { val: "'asd'" }, a: { val: 2 } },
        where: [{ ref: ['a'] }, '<', { val: 9 }],
      },
    }
    const { sql, values } = cqn2sql(cqnUpdate)
    expect({ sql, values }).toMatchSnapshot()
  })

  test('test with setting a value to null', () => {
    const cqnUpdate = {
      UPDATE: {
        entity: { ref: ['Foo2'] },
        with: { ID: { val: 1 }, name: { val: null }, a: { val: 2 } },
        where: [{ ref: ['a'] }, '<', { val: 9 }],
      },
    }

    const { sql, values } = cqn2sql(cqnUpdate)
    expect({ sql, values }).toMatchSnapshot()
  })

  test('test with entity and values with operators', () => {
    const cqnUpdate = {
      UPDATE: {
        entity: 'Foo2',
        with: {
          ID: { val: 42 },
          name: { val: "'asd'" },
          a: { xpr: [{ ref: ['a'] }, '-', { val: 1 }] },
          count: { func: 'count', args: ['*'] },
        },
      },
    }
    const { sql, values } = cqn2sql(cqnUpdate)
    expect({ sql, values }).toMatchSnapshot()
  })

  //REVISIT aliasing with columns doesn't work
  test('data alone still works', () => {
    const cqnUpdate = {
      UPDATE: {
        entity: { ref: ['Foo2'] },
        data: {
          ID: 1,
          name: undefined,
          a: null,
        },
      },
    }
    const { sql, values } = cqn2sql(cqnUpdate)
    expect({ sql, values }).toMatchSnapshot()
  })

  test('virtual and non-existing filtered out from data', () => {
    const cqnUpdate = {
      UPDATE: {
        entity: { ref: ['Foo2'] },
        data: {
          ID: 1,
          something: 'bla',
          foo: null,
        },
      },
    }
    const { sql, values } = cqn2sql(cqnUpdate)
    expect({ sql, values }).toMatchSnapshot()
  })

  test('set enhances data', () => {
    const cqnUpdate = {
      UPDATE: {
        entity: { ref: ['Foo2'] },
        with: { ID: { val: 1 }, name: { val: "'asd'" } },
        data: { a: 2 },
      },
    }

    const { sql, values } = cqn2sql(cqnUpdate)
    expect({ sql, values }).toMatchSnapshot()
  })

  test('virtual and non-existing fields filtered out from with', () => {
    const cqnUpdate = {
      UPDATE: {
        entity: { ref: ['Foo2'] },
        with: { ID: { val: 1 }, name: { val: "'asd'" }, something: { val: 'bla' } /* foo: {ref: 'Foo'} */ },
      },
    }

    const { sql, values } = cqn2sql(cqnUpdate)
    expect({ sql, values }).toMatchSnapshot()
  })

  test('set overwrites data', () => {
    const cqnUpdate = {
      UPDATE: {
        entity: 'Foo2',
        with: { ID: { val: 1 }, name: { val: "'asd'" }, a: { val: 6 } },
        data: { a: 2 },
      },
    }

    const { sql, values } = cqn2sql(cqnUpdate)
    expect({ sql, values }).toMatchSnapshot()
  })

  // TODO change to native sql -> not really useful to test here
  test.skip('test with subselect - sflight example', () => {
    const qlUpdate = UPDATE(`Travel`).with({
      TotalPrice: CXL`coalesce (BookingFee, 0) + ${SELECT`coalesce (sum (FlightPrice + ${SELECT`coalesce (sum (Price),0)`.from(
        `BookingSupplement`,
      ).where`to_Booking_BookingUUID = BookingUUID`}),0)`.from(`Booking`).where`to_Travel_TravelUUID = TravelUUID`}`,
    })
    const { sql, values } = cqn2sql(qlUpdate)
    expect({ sql, values }).toMatchSnapshot()
  })
})
