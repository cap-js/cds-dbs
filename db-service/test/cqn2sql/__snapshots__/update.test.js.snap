// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`.update data alone still works 1`] = `
{
  "sql": "UPDATE Foo2 AS Foo2 SET ID=1,name=NULL,a=NULL",
  "values": [],
}
`;

exports[`.update set enhances data 1`] = `
{
  "sql": "UPDATE Foo2 AS Foo2 SET a=2,ID=1,name=?",
  "values": [
    "'asd'",
  ],
}
`;

exports[`.update set overwrites data 1`] = `
{
  "sql": "UPDATE Foo2 AS Foo2 SET a=2,ID=1,name=?,a=6",
  "values": [
    "'asd'",
  ],
}
`;

exports[`.update test with entity and values with operators 1`] = `
{
  "sql": "UPDATE Foo2 AS Foo2 SET ID=42,name=?,a=Foo2.a - 1",
  "values": [
    "'asd'",
  ],
}
`;

exports[`.update test with entity of type string 1`] = `
{
  "sql": "UPDATE Foo2 AS Foo2 SET ID=1,name=?,a=2",
  "values": [
    "'asd'",
  ],
}
`;

exports[`.update test with entity of type string and where clause 1`] = `
{
  "sql": "UPDATE Foo2 AS Foo2 SET ID=1,name=?,a=2 WHERE Foo2.a < 9",
  "values": [
    "'asd'",
  ],
}
`;

exports[`.update test with setting a value to null 1`] = `
{
  "sql": "UPDATE Foo2 AS Foo2 SET ID=1,name=NULL,a=2 WHERE Foo2.a < 9",
  "values": [],
}
`;

exports[`.update test with subselect - sflight example 1`] = `
{
  "sql": "UPDATE Travel AS Travel SET TotalPrice=coalesce(BookingFee,0) + (SELECT coalesce(sum(FlightPrice + (SELECT coalesce(sum(Price),0) as coalesce FROM BookingSupplement WHERE to_Booking_BookingUUID = BookingUUID)),0) as coalesce FROM Booking WHERE to_Travel_TravelUUID = TravelUUID)",
  "values": [],
}
`;

exports[`.update virtual and non-existing fields filtered out from with 1`] = `
{
  "sql": "UPDATE Foo2 AS Foo2 SET ID=1,name=?",
  "values": [
    "'asd'",
  ],
}
`;

exports[`.update virtual and non-existing filtered out from data 1`] = `
{
  "sql": "UPDATE Foo2 AS Foo2 SET ID=1",
  "values": [],
}
`;
